pkg/kubelet/images/image_manager.go

```diff
@@ -64,27 +87,45 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
 	} else {
 		puller = newParallelImagePuller(imageService, maxParallelImagePulls)
 	}
+
 	return &imageManager{
-		recorder:               recorder,
-		imageService:           imageService,
-		backOff:                imageBackOff,
-		puller:                 puller,
-		podPullingTimeRecorder: podPullingTimeRecorder,
+		recorder:                     recorder,
+		imageService:                 imageService,
+		backOff:                      imageBackOff,
+		puller:                       puller,
+		podPullingTimeRecorder:       podPullingTimeRecorder,
+		keyring:                      keyring,
+		ensureSecretPulledImages:     make(map[string]*ImagePullInfo),
+		pullImageSecretRecheckPeriod: pullImageSecretRecheckPeriod,
+		pullImageSecretRecheck:       pullImageSecretRecheck,
 	}
 }
 
 // shouldPullImage returns whether we should pull an image according to
 // the presence and pull policy of the image.
-func shouldPullImage(container *v1.Container, imagePresent bool) bool {
-	if container.ImagePullPolicy == v1.PullNever {
+func shouldPullImage(container *v1.Container, imagePresent, pulledBySecret, ensuredBySecret, pullImageSecretRecheck bool) bool {
+	switch container.ImagePullPolicy {
+	case v1.PullNever:
```

Comment:
we still need to do some checking here. if it's not in the cache then we need to fail because the image may be present on the node but we're not allowed to use it

Reply:
ah nevermind, this function is just answering the question "should we pull this image" which the answer for pull never is always no

Reply:
This scenario has been taken care before calling the `shouldPullImage` i.e. in the `isEnsuredBySecret` function.